src/cli/self_update.rs

```diff
@@ -111,71 +114,19 @@ async fn latest_version() -> miette::Result<Version> {
     }
 }
 
-pub async fn execute(args: Args) -> miette::Result<()> {
-    // Get the target version, without 'v' prefix
-    let target_version = match &args.version {
-        Some(version) => version,
-        None => &latest_version().await?,
-    };
-    // Get the current version of the pixi binary
-    let current_version = Version::from_str(consts::PIXI_VERSION).into_diagnostic()?;
-
-    // Stop here if the target version is the same as the current version
-    if *target_version == current_version {
-        eprintln!(
-            "{}pixi is already up-to-date (version {})",
-            console::style(console::Emoji("✔ ", "")).green(),
-            current_version
-        );
-        return Ok(());
-    }
-
-    let action = if *target_version < current_version {
-        if args.version.is_none() {
-            // Ask if --version was not passed
-            let confirmation = dialoguer::Confirm::new()
-                .with_prompt(format!(
-                    "\nCurrent version ({}) is more recent than remote ({}). Do you want to downgrade?",
-                    current_version, target_version
-                ))
-                .default(false)
-                .show_default(true)
-                .interact()
-                .into_diagnostic()?;
-            if !confirmation {
-                return Ok(());
-            };
-        };
-        "downgraded"
-    } else {
-        "updated"
-    };
-
-    eprintln!(
-        "{}Pixi will be {} from {} to {}",
-        console::style(console::Emoji("✔ ", "")).green(),
-        action,
-        current_version,
-        target_version
-    );
-
-    // Get the name of the binary to download and install based on the current platform
-    let archive_name = default_archive_name()
-        .expect("Could not find the default archive name for the current platform");
-
-    let download_url = format!(
-        "{}/download/v{}/{}",
-        consts::RELEASES_URL,
-        target_version,
-        archive_name
-    );
+/// Downloads the target of an URL into a temporary file.
+async fn download<U>(url: U) -> miette::Result<NamedTempFile>
```

Comment:
maybe we can accept here f: W where W: AsyncTrait?

in this case, we can use this function with providing our own file and the responsibility to setup it will be on caller side